### ISSUE TITLE
[gql_build] interface and union types

### ISSUE THREAD

**Issue by smkhalsa** [gql-dart/gql#38]
@klavs have you thought about how to implement interface and union types?

**Comment by klavs** [comment]
I was thinking that a user shouldn't be able to access any union fields (except `__typename`) and any fields not on the interface type before casting to the concrete type.

The problem with casting is that it would be hard to use Dart interfaces because the generated class names are not supposed to be used directly.

The only alternative I've thought of would be to add special field `$asSomeConcreteType` in the generated class which would return `null` if `__typename != "SomeConcreteType"` or the a non-null data class.
It might be useful to include `$isSomeConcreteType` which returns boolean.

**Comment by smkhalsa** [comment]
> I was thinking that a user shouldn't be able to access any union fields (except __typename) and any fields not on the interface type before casting to the concrete type.

I agree.

> The problem with casting is that it would be hard to use Dart interfaces because the generated class names are not supposed to be used directly.

> The only alternative I've thought of would be to add special field $asSomeConcreteType in the generated class which would return null if __typename != "SomeConcreteType" or the a non-null data class.
It might be useful to include $isSomeConcreteType which returns boolean.

I was thinking that we could generate a concrete type for the graphql interface type, have the subtypes implement that interface, and instantiate the appropriate subtype in the supertype constructor based on the `__typename` field.

So for example, 

```dart
// GraphQL Interface type
class $Audio {
  final Map<String, dynamic> data;

  String get id => data['id'];
  // ... other getters

  $Audio._withData(this.data);

  factory $Audio(Map<String, dynamic> data) {
    switch (data['__typename']) {
      case 'Song':
        return $Audio$asSong(data);
      case 'Podcast':
        return $Audio$asPodcast(data);
      default:
        return $Audio._withData(data);
    }
  }
}

class $Audio$asSong implements $Audio {
  final Map<String, dynamic> data;

  // ... $Audio getters
  // ... Song-specific getters

  $Audio$asSong(this.data);
}

class $Audio$asPodcast implements $Audio {
  final Map<String, dynamic> data;

  // ... $Audio getters
  // ... Podcast-specific getters

  $Audio$asPodcast(this.data);
}
```

**Comment by smkhalsa** [comment]
A GraphQL union could have a similar implementation, but the super class would only include a typename field:

```dart
// GraphQL Union type
class $Song$Podcast {
  final Map<String, dynamic> data;

  String get $typename => data['__typename'];

  $Song$Podcast._withData(this.data);

  factory $Song$Podcast(Map<String, dynamic> data) {
    switch (data['__typename']) {
      case 'Song':
        return $$Song$Podcast$asSong(data);
      case 'Podcast':
        return $Song$Podcast$asPodcast(data);
      default:
        return $Song$Podcast._withData(data);
    }
  }
}

class $Song$Podcast$asSong implements $Song$Podcast {
  final Map<String, dynamic> data;

  // ... Song-specific getters

  $Song$Podcast$asSong(this.data);
}

class $Song$Podcast$asPodcast implements $Song$Podcast {
  final Map<String, dynamic> data;

  // ... Podcast-specific getters

  $Song$Podcast$asPodcast(this.data);
}
```

**Comment by klavs** [comment]
We shouldn't think of object types as classes. That's incorrect. Object types are just templates defining how a concrete class may look like. The only things that are concrete classes are selection sets defined by operations or fragments.

Another important thing is that using N interfaces in your document, should be as simple as using 1 interface.

**Comment by smkhalsa** [comment]
> We shouldn't think of object types as classes. That's incorrect. Object types are just templates defining how a concrete class may look like. The only things that are concrete classes are selection sets defined by operations or fragments.

I think I agree with you. In the interface example above, the `$Audio` class was meant to be a representation of the shared selection set fields defined by an operation or fragment. The class name probably should have been `$[MyOperationName]$Audio`.

In other words, each operation that uses an inline fragment would have its own `$[MyOperationName]$Audio` generated.

For example, assuming we have this schema:

```graphql
interface Audio {
  id: String!
  name: String!
  fileUrl: String!
}

type Song implements Audio {
  id: String!
  name: String!
  fileUrl: String!
  album: Album!
  artist: Artist!
}

type Album {
  id: String!
  name: String!
}

type Artist {
  id: String!
  name: String!
}

type Query {
  audio(id: String!): Audio
}
```

... and define this query:

```graphql
query GetAudio($id: String!) {
  audio(id: $id) {
    id
    name
    fileUrl
    ... on Song {
      album {
        id
        name
      }
      artist {
        id
        name
      }
    }
  }
}
```

The `data_builder` would generate a concrete `$GetAudio$Audio` class as well as a `$GetAudio$Audio$asSong` class that implements the `$GetAudio$Audio` interface.

This way the user code can consume any object that is an instance of `$GetAudio$Audio` (in their audio player code, for example).

**Comment by smkhalsa** [comment]
@klavs after digging into this a bit more, I realized that you're right that we don't need to create a separate dart class for the `interface` or `union` to accomplish my example above.

For the inline fragment example above, we could simply build a `$GetAudio` class as well as a subclass`$GetAudio$asSong`.

I've implemented this [here](https://github.com/gql-dart/gql/pull/32/commits/be73ff849ec27c02bf57329a070f577a22002dfa), so I'll close this issue.
